### ISSUE TITLE
Error Airbrake: performance notifier closed causes deadlock

### ISSUE THREAD

**Issue by hienvd** [airbrake/airbrake-ruby#524]
* Airbrake version: 9.5.5
* Ruby version: 2.6.3
* Rails version: 5.2.3

#### Description
I encountered a problem that the performance notifier was closed that caused this error:
`Airbrake::Error (attempted to build deadlock; recursive locking with closed Airbrake instance)` 

When trace in log, I saw some errors in this order:
```bash
Airbrake: HTTP error: Failed to open TCP connection to api.airbrake.io:443 (detected a network failure: Empty response received from the server.)
Airbrake: thread pool closed
Airbrake: performance notifier closed
Airbrake::Error (attempted to build deadlock; recursive locking with closed Airbrake instance)
```

There were bunch of this *deadlock* error that made my app didn't work normally.
Do you have any ideas?

**Comment by kyrylo** [comment]
@hienvd I applied a "fix" in https://github.com/airbrake/airbrake-ruby/pull/525. There are at least 2 problems here. One is the confusing error message (fixed by #525), the other one is the attempt to send a notice after Airbrake was closed.

FYI performance notifier is probably not responsible for the deadlock because the error you saw is from the notice notifier. The error itself could be coming from anywhere within your app. In simple terms, this is what happens:

```
[3] pry(main)> Airbrake.close
D, [2019-12-16T19:37:18.061480 #56101] DEBUG -- : **Airbrake: thread pool closed
=> nil
[4] pry(main)> Airbrake.notify('asd')
W, [2019-12-16T19:37:25.924571 #56101]  WARN -- : **Airbrake: falling back to sync delivery because there are no running async workers
Airbrake::Error: attempted to build asd with closed Airbrake instance
```

The exception that you try to send is likely a `ThreadError`, and it's hard to tell whose fault that is.
Do you have a full backtrace when `Airbrake::Error` is raised? Maybe you could increase your logger level to DEBUG?

**Comment by hienvd** [comment]
Thank @kyrylo for your investigation.

For the main problem, I don't have a full backtrace, but I think this might help:

```bash
D, [2019-12-12T18:08:07.746914 #48333] DEBUG -- : [4b4ffb37-c3cb-4721-ab78-fa0df59173be]   Job Load (0.5ms)  SELECT  "jobs".* FROM "jobs" WHERE "jobs"."deleted_at" IS NULL AND "jobs"."slug" = $1 LIMIT $2  [["slug", "abcd-5527"], ["LIMIT", 1]]
D, [2019-12-12T18:08:07.748852 #48298] DEBUG -- : [d29d6632-3fe6-4977-a5f3-1956be99b95e]    (0.3ms)  BEGIN
F, [2019-12-12T18:08:07.749596 #48333] FATAL -- : [4b4ffb37-c3cb-4721-ab78-fa0df59173be]   
F, [2019-12-12T18:08:07.749665 #48333] FATAL -- : [4b4ffb37-c3cb-4721-ab78-fa0df59173be] Airbrake::Error (attempted to build deadlock; recursive locking with closed Airbrake instance):
F, [2019-12-12T18:08:07.749705 #48333] FATAL -- : [4b4ffb37-c3cb-4721-ab78-fa0df59173be]   
F, [2019-12-12T18:08:07.749771 #48333] FATAL -- : [4b4ffb37-c3cb-4721-ab78-fa0df59173be] airbrake-ruby (4.8.0) lib/airbrake-ruby/notice_notifier.rb:57:in `build_notice'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] airbrake-ruby (4.8.0) lib/airbrake-ruby.rb:242:in `build_notice'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] airbrake (9.5.5) lib/airbrake/rack/middleware.rb:66:in `notify_airbrake'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] airbrake (9.5.5) lib/airbrake/rack/middleware.rb:34:in `rescue in call!'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] airbrake (9.5.5) lib/airbrake/rack/middleware.rb:31:in `call!'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] airbrake (9.5.5) lib/airbrake/rack/middleware.rb:21:in `call'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] actionpack (5.2.3) lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] actionpack (5.2.3) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] newrelic_rpm (6.7.0.359) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] lograge (0.10.0) lib/lograge/rails_ext/rack/logger.rb:15:in `call_app'
[4b4ffb37-c3cb-4721-ab78-fa0df59173be] railties (5.2.3) lib/rails/rack/logger.rb:26:in `block in call'
```

**Comment by kyrylo** [comment]
Thanks for the extra info. Unfortunately, I still struggle to see why this is happening. We do not close Airbrake anywhere else but here (on `at_exit`):

https://github.com/airbrake/airbrake/blob/f65eaa431e16bf15ae6c6b457ea62d0ba165fb2a/lib/airbrake.rb#L26-L30

I can see two scenarios:

* your program wants to exit and our `at_exit` hook gets fired but then something else rescues the process so it keeps running
* you or a third party library calls `Airbrake.close`

You mention that it happens when a request hits your controller action. Can you describe what the action does? I assume it's a simple `index` action that lists jobs.

**Comment by hienvd** [comment]
Thank you for your information.
That is a normal controller/action, as you can see in the backtrace above.  
I didn't close Airbrake, I'm pretty sure that no 3rd party library do this. 

As described in the issue, I noticed a `HTTP error: Failed to open TCP connection to api.airbrake.io:443` first. Could this cause the problem?

**Comment by hienvd** [comment]
Hi @kyrylo ,
Also, with the first problem, how can I apply your fix in #525?
I do not see any new release version of `airbrake` or `airbrake-ruby`.

**Comment by kyrylo** [comment]
> Also, with the first problem, how can I apply your fix in #525?

Correct, it hasn't been released yet. If you really want it, you could depend on the commit hash in your Gemfile.

> As described in the issue, I noticed a HTTP error: Failed to open TCP connection to api.airbrake.io:443 first. Could this cause the problem?

I set up a test Rails app and an action that always raises (and therefore sends exceptions to Airbrake). I disabled network connection and hit my localhost with a request. When `Airbrake.notify` is called, I see this error in the logs:

```
ERROR -- : **Airbrake: HTTP error: Failed to open TCP connection to api.airbrake.io:443 (getaddrinfo: nodename nor servname provided, or not known)
```

This is expected. However, I cannot replicate the issue you mentioned. Thread pools still continue working, so I can spam as many requests as I want.

If you do the same for your app can you see the error locally?

**Comment by hienvd** [comment]
Thank you for your support @kyrylo . I really appreciate it.

I tested locally, sending request to an action that always raises while network connection being disabled. I got the same error, but it didn't cause the deadlock error.

I have not had that `deadlock` error so far, but it is weird and I don't know if it will happen in the future or not. I have encountered this error at least 2 times in the last 2 weeks

**Comment by hienvd** [comment]
Hi @kyrylo ,
I don't know if there is any relation between my issue and this one #517 (which was fixed in #520 )? 
I'm using `airbrake 9.5.5`, depended on `airbrake-ruby 4.8.0`

**Comment by kyrylo** [comment]
I don't know how to reproduce the error and it's hard to make a sensible action here. I would guess that it shouldn't matter, but if you can upgrade and monitor this for a week or two, it'd be a great help.

**Comment by hienvd** [comment]
I've just upgrade airbrake-ruby to the latest version.
Thank again for your help!

**Comment by kyrylo** [comment]
I just released v4.11.0, which includes the fix for the error message. You might want to pull it.

**Comment by kyrylo** [comment]
Hey @hienvd, have you been able to monitor the new version of the library?
